resource/cloudflare_zone_cache_reserve: add a new resource type

kwilczynski opened the pull request:
Add a new resource that can be used to manage a [Cache Reserve](https://developers.cloudflare.com/cache/advanced-configuration/cache-reserve/) feature of a given zone.

While at it, add a new data source to complete the newly added resource so that the current status of the Cache Reserve feature can be read for a given zone.

Closes: [#2420 ](https://github.com/cloudflare/terraform-provider-cloudflare/issues/2420)

Related to:

- https://github.com/cloudflare/cloudflare-docs/pull/8561
- https://github.com/cloudflare/cloudflare-go/issues/1277

More details can be found at:

- Cloudflare Documentation:
  - [Tiered Cache](https://developers.cloudflare.com/cache/how-to/tiered-cache)
  - [Cache Reserve](https://developers.cloudflare.com/cache/advanced-configuration/cache-reserve)
- Cloudflare Blog:
  - [Introducing: Smarter Tiered Cache Topology Generation](https://blog.cloudflare.com/introducing-smarter-tiered-cache-topology-generation)
  - [Introducing Cache Reserve: massively extending Cloudflare’s cache](https://blog.cloudflare.com/introducing-cache-reserve)

kwilczynski commented:
Unit tests:

```
$ go test ./... -count 1 -parallel 1 -race
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/internal/acctest	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/internal/consts	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/expanders	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/flatteners	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/modifiers/defaults	[no test files]
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/provider	0.040s [no tests to run]
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/service/example	0.036s [no tests to run]
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/service/r2_bucket	0.033s
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/service/rulesets	0.041s
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/framework/service/turnstile	0.031s
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider	0.284s
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/utils	0.080s
```

Acceptance tests:

```
$ TF_ACC=1 go test ./internal/sdkv2provider/ -v -count 1 -run '^TestAcc(.+)?CloudflareZoneCacheReserve' -timeout 120m -parallel 1
=== RUN   TestAccDataCloudflareZoneCacheReserve_Simple
=== PAUSE TestAccDataCloudflareZoneCacheReserve_Simple
=== RUN   TestAccDataCloudflareZoneCacheReserve_Error
=== PAUSE TestAccDataCloudflareZoneCacheReserve_Error
=== RUN   TestAccCloudflareZoneCacheReserve_Import
=== PAUSE TestAccCloudflareZoneCacheReserve_Import
=== RUN   TestAccCloudflareZoneCacheReserve_Simple
=== PAUSE TestAccCloudflareZoneCacheReserve_Simple
=== RUN   TestAccCloudflareZoneCacheReserve_Error
=== PAUSE TestAccCloudflareZoneCacheReserve_Error
=== CONT  TestAccDataCloudflareZoneCacheReserve_Simple
--- PASS: TestAccDataCloudflareZoneCacheReserve_Simple (4.19s)
=== CONT  TestAccCloudflareZoneCacheReserve_Simple
--- PASS: TestAccCloudflareZoneCacheReserve_Simple (5.08s)
=== CONT  TestAccCloudflareZoneCacheReserve_Error
--- PASS: TestAccCloudflareZoneCacheReserve_Error (0.13s)
=== CONT  TestAccCloudflareZoneCacheReserve_Import
--- PASS: TestAccCloudflareZoneCacheReserve_Import (4.11s)
=== CONT  TestAccDataCloudflareZoneCacheReserve_Error
--- PASS: TestAccDataCloudflareZoneCacheReserve_Error (0.12s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider	13.631s
```

kwilczynski commented on internal/sdkv2provider/schema_cloudflare_zone_cache_reserve.go:
```diff
@@ -0,0 +1,29 @@
+				return
+			},
+		},
+		"enabled": {
```
**To the reviewer:**

I made this to be a **required** attribute on purpose, as this has both a monetary value attached to it and performance side-effects when enabled (or disabled), especially since it requires a dedicated pricing plan to be enabled, and as such user has to explicitly choose to enable or disable this feature of their zone.

Thus, no default setting is being offered here.

jacobbednarz commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 
+FEATURES:
+
+* **New Data Source:** `cloudflare_zone_cache_reserve` ([#2642](https://github.com/cloudflare/terraform-provider-cloudflare/issues/2642))
```
this file is automatically updated on merge so you don't need to commit this.

kwilczynski replied:
Got it. I wasn't sure how much of a changelog would be required. Will fix.

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on examples/resources/cloudflare_zone_cache_reserve/resource.tf:
```diff
@@ -0,0 +1,5 @@
+# Enable the Cache Reserve support for a given zone.
```
```suggestion
```

kwilczynski replied:
Will fix.

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on examples/resources/cloudflare_zone_cache_reserve/resource.tf:
```diff
@@ -0,0 +1,5 @@
+# Enable the Cache Reserve support for a given zone.
+resource "cloudflare_zone_cache_variants" "example" {
```
```suggestion
resource "cloudflare_zone_cache_reserve" "example" {
```

kwilczynski replied:
Doh! Well spotted! Will fix!

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on internal/sdkv2provider/resource_cloudflare_zone_cache_reserve.go:
```diff
@@ -0,0 +1,138 @@
+
+			Note: Using Cache Reserve without Tiered Cache is not recommended.
+
+			Requires Cache Reserve subscription.
```
we can omit this as we don't use treat the terraform docs as source of what subscriptions are required for a product to work.

```suggestion
```

kwilczynski replied:
Got it. Will fix. My reasoning was to ensure that users are aware that there are some requirements for this to work - otherwise, they will get a cryptic error message back, as I did. :smile:

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on internal/sdkv2provider/data_source_zone_cache_reserve.go:
```diff
@@ -0,0 +1,71 @@
+		Description: heredoc.Doc(`
+			Provides a Cloudflare data source to look up Cache Reserve
+			status for a given zone.
+
+			Requires Cache Reserve subscription.
```
```suggestion
```

kwilczynski replied:
Will fix.

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on internal/sdkv2provider/data_source_zone_cache_reserve.go:
```diff
@@ -0,0 +1,71 @@
+	d.Set(consts.ZoneIDSchemaKey, zoneID)
+	d.Set("enabled", output.Value == cacheReserveEnabled)
+
+	d.SetId(stringChecksum(output.ModifiedOn.String()))
```
this may be a little too high entropy for consistent use. could we set the zone_id here instead as the anchor?

kwilczynski replied:
Both change at the same time, but updating the **enabled** attribute only looked better, indeed. Will change.

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on internal/sdkv2provider/resource_cloudflare_zone_cache_reserve.go:
```diff
@@ -0,0 +1,138 @@
+	if err != nil {
+		return diag.Errorf("unable to delete Cache Reserve for zone %q: %s", zoneID, err)
+	}
+
```
we should call `d.SetId("")` after this update too in order to remove it from the state file.

kwilczynski replied:
On the error path, we probably don't want to call it as it would orphan the resource should we fail to delete it. On the happy path, so to speak, there should be no need to call this explicitly nowadays as it would be called automatically for us when this callback returns without an error.

I would add it in the past, but it seems now this pattern is not desirable per:

- Developer Guide: [Implement delete](https://developer.hashicorp.com/terraform/tutorials/providers/provider-delete)
- https://github.com/hashicorp/terraform-provider-aws/issues/18390
- https://github.com/hashicorp/terraform-provider-aws/pull/4415

Which is why I didn't add it here to the delete callback. Thoughts? Do you still want it added?

kwilczynski replied:
However, I will add handling the error should the zone no longer be available to make deletion more resilient.

jacobbednarz replied:
> On the happy path, so to speak, there should be no need to call this explicitly nowadays as it would be called automatically for us when this callback returns without an error.

this is a nice DX improvement. thanks for the heads up!

jacobbednarz commented on internal/sdkv2provider/validators.go:
```diff
@@ -74,3 +81,12 @@ func validateURL(v interface{}, k string) (s []string, errors []error) {
 }
+
+// validateZoneID ensures that the given Zone ID is valid.
+func validateZoneID(value string) error {
```
this is a nice validator but a little overkill. we generally don't enforce this as at it's core, the value is just a hex string and no way to actually determine if it is a zone, account or other external identifier.

kwilczynski replied:
Do you want it removed? I added it primarily for the import feature, as when you provide something that does not conform to the known format. Terraform terminates with an error and shows a message claiming the provider will be blamed for an error.

I can either drop this or move the validation into the import function. Thoughts?

kwilczynski replied:
Done. Thank you!

kwilczynski commented on templates/data-sources/zone_cache_reserve.md.tmpl:
```diff
@@ -0,0 +1,23 @@
+
+# {{.Name}} ({{.Type}})
+
+~> Requires Cache Reserve subscription.
```
This will be removed as per the feedback.

kwilczynski replied:
Done. Thank you!

kwilczynski commented on templates/resources/zone_cache_reserve.md.tmpl:
```diff
@@ -0,0 +1,27 @@
+
+# {{.Name}} ({{.Type}})
+
+~> Requires Cache Reserve subscription.
```
This will be removed as per the feedback.

kwilczynski replied:
Done. Thank you!

jacobbednarz commented on templates/data-sources/zone_cache_reserve.md.tmpl:
do we need this file? any reason why we can't use the generic template? the description in here should be using `{{ .Description | trimspace }}` instead which will allow us to keep the templates the same.

kwilczynski replied:
I wasn't sure how to add a link to the Cache Reserve documentation site without a template. Unless you believe it's not needed, then we can remove it.

jacobbednarz commented on templates/resources/zone_cache_reserve.md.tmpl:
similar to the other template file; can we do without this? we don't usually include things like which subscriptions are needed in the terraform docs.

kwilczynski replied:
I only added it for the sake of being able to add a link to the documentation.

I can remove this if you believe it's OK not to link users to the documentation in the resource description, which may be too much given the documentation generation automation that aims to keep everything consistent.

Thoughts?

kwilczynski replied:
Done. Thank you!

kwilczynski commented on internal/sdkv2provider/resource_cloudflare_zone_cache_reserve.go:
```diff
@@ -0,0 +1,138 @@
+	output, err := client.GetCacheReserve(ctx, cloudflare.ZoneIdentifier(zoneID), params)
+	if err != nil {
+		var notFoundError *cloudflare.NotFoundError
+		if errors.As(err, &notFoundError) {
```
@jacobbednarz, would this ever be a problem for Cache Reserve?

While testing the code during development, when I selected a non-existent zone ID, albeit still a valid format, I got a permissions/access error rather than a 404. I am unsure how this is implemented on the backend side, though.

That said, I didn't check a non-sensical ID such as "this is a test" to something along these lines. I will do it later to see what will happen.

kwilczynski replied:
I will answer my question: **GetCacheReserve()** returns a **404**, while **UpdateCacheReserve()** returns a different type of error message. I also ensured that 404 is handled correctly in the **Delete** callback.

kwilczynski commented:
Updated unit and acceptance tests following addressing review feedback: https://github.com/cloudflare/terraform-provider-cloudflare/pull/2642#issuecomment-1659014946

jacobbednarz commented:
thanks @kwilczynski, appreciate the effort getting this one over the line 👏 🏆
